Move subresources tests under compute pkg and mark conformance tests

fossedihelm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does

`subresource_test.go` file was too generic and was holding tests that are more vmi/vm lifecycle related.
In addition, a lot of tests can be converted to unit tests.
This PR aims to refactor the subresource_test file doing the following:
- e2e to unit test conversion when possible.
- Extract tests into more specific context (expandspec, openapi, rbac, vmi/vm lifecycle).
- Less vmi setup on BeforeEach block.
- Mark conformance tests following https://github.com/kubevirt/kubevirt/blob/main/docs/conformance.md

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

xpivarc commented on tests/compute/subresources/rbac.go:
```diff
@@ -47,25 +48,25 @@ var _ = compute.SIGDescribe("[rfe_id:1195][crit:medium][vendor:cnv-qe@redhat.com
 		})
 
-		It("[test_id:3170]should allow access to vm subresource endpoint", func() {
+		It("[test_id:3170]should allow access to vm subresource endpoint", decorators.Conformance, func() {
```
Similar as to the rbac tests, I am not sure this is good candidate

xpivarc replied:
@fossedihelm You seem to agree but I see that this is still included. What do you plan to do with this?

fossedihelm replied:
@xpivarc Sorry, I probably missed that during the rebase.
Dropped the commit.

xpivarc commented on tests/compute/subresources/vm.go:
```diff
@@ -52,7 +52,7 @@ var _ = compute.SIGDescribe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com
 
 	Context("with a restart endpoint", func() {
-		DescribeTable("should return an error when VM is not running", func(errMsg string, opts ...libvmi.VMOption) {
+		DescribeTable("should return an error when VM is not running", decorators.Conformance, func(errMsg string, opts ...libvmi.VMOption) {
```
This looks like unit test to me.

xpivarc commented on tests/compute/subresources/vm.go:
```diff
@@ -66,7 +66,7 @@ var _ = compute.SIGDescribe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com
 		)
 
-		It("[test_id:2265][posneg:negative] should return an error when VM has not been found but VMI is running", func() {
+		It("[test_id:2265][posneg:negative] should return an error when VM has not been found but VMI is running", decorators.Conformance, func() {
```
Same as above

xpivarc commented on tests/compute/subresources/vm.go:
```diff
@@ -96,7 +96,7 @@ var _ = compute.SIGDescribe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com
 	})
 
-	It("[test_id:1530]should stop a VirtualMachine only once", func() {
+	It("[test_id:1530]should stop a VirtualMachine only once", decorators.Conformance, func() {
```
Same as above

xpivarc commented on tests/compute/subresources/pausing.go:
```diff
@@ -156,7 +157,7 @@ var _ = compute.SIGDescribe("[rfe_id:3064][crit:medium][vendor:cnv-qe@redhat.com
 		})
 
-		It("[test_id:3081]should gracefully handle pausing the VM again", func() {
+		It("[test_id:3081]should gracefully handle pausing the VM again", decorators.Conformance, func() {
```
Could be merged to different test

xpivarc commented on tests/compute/subresources/pausing.go:
```diff
@@ -265,7 +266,7 @@ var _ = compute.SIGDescribe("[rfe_id:3064][crit:medium][vendor:cnv-qe@redhat.com
 		})
 
-		It("[test_id:3083]should connect to serial console", func() {
+		It("[test_id:3083]should connect to serial console", decorators.Conformance, func() {
```
Leaving comment here but it generally hold for the whole commit. 
1. Few of tests are written in a way that assume they are cheap and it is better to write one more test for one more assert rather than writing the assert to existing test.
2. Console and VNC should have separate tests/files

fossedihelm commented on tests/compute/subresources/rbac.go:
```diff
@@ -0,0 +1,123 @@
+	"kubevirt.io/kubevirt/tests/testsuite"
+)
+
+var _ = compute.SIGDescribe("[rfe_id:1195][crit:medium][vendor:cnv-qe@redhat.com][level:component] Rbac authorization", func() {
```
As per offline discussion, I will address this in a separate PR, to not enlarge the scope of the PR too much

fossedihelm commented on tests/compute/subresources/rbac.go:
```diff
@@ -0,0 +1,123 @@
+	secret, err := virtCli.CoreV1().Secrets(testsuite.GetTestNamespace(nil)).Get(context.Background(), saName, metav1.GetOptions{})
+	Expect(err).ToNot(HaveOccurred())
+
+	token, ok := secret.Data["token"]
```
In `tests/serviceaccount.go` we also create the secret alongside the service account.
```
	_, err = virtCli.CoreV1().Secrets(GetTestNamespace(nil)).Create(context.Background(), &secret, metav1.CreateOptions{})
	if !k8serrors.IsAlreadyExists(err) {
		util.PanicOnError(err)
	}
```

fossedihelm commented:
@xpivarc Comments should have been addresses, PTAL

fossedihelm commented:
/retest

xpivarc commented on tests/compute/subresources/vmi.go:
```diff
@@ -75,14 +76,14 @@ var _ = compute.SIGDescribe("VirtualMachineInstance subresource", func() {
 		})
 
-		It("[test_id:7476]Freeze without guest agent", func() {
+		It("[test_id:7476]Freeze without guest agent", decorators.Conformance, func() {
```
This is too simple and should be unit test.

fossedihelm replied:
Honestly I find it hard to convert it to a unit test.
Currently, there is no explicit logic that blocks the freeze for a vmi without guestagentconnected.
But it relies on the fact that the put request on the virt-handler side fails.
I can convert to unit test mocking the failing of virt-handler connection, but honestly it is not strictly related to guest agent connected. Should it works for you?

xpivarc commented on tests/compute/subresources/vmi.go:
```diff
@@ -75,14 +76,14 @@ var _ = compute.SIGDescribe("VirtualMachineInstance subresource", func() {
 		})
 
-		It("[test_id:7477]Unfreeze without guest agent", func() {
+		It("[test_id:7477]Unfreeze without guest agent", decorators.Conformance, func() {
```
This is too simple and should be unit test.

xpivarc commented on tests/compute/subresources/vmi.go:
```diff
@@ -122,7 +123,7 @@ var _ = compute.SIGDescribe("VirtualMachineInstance subresource", func() {
 		}
 
-		It("[test_id:7479]Freeze Unfreeze should succeed", func() {
+		It("[test_id:7479]Freeze Unfreeze should succeed", decorators.Conformance, func() {
```
I would hold these tests to higher standard. I would like to see the setup here.

xpivarc commented on tests/compute/subresources/vmi.go:
```diff
@@ -181,7 +182,7 @@ var _ = compute.SIGDescribe("VirtualMachineInstance subresource", func() {
 		})
 
-		It("soft reboot vmi with ACPI feature enabled should succeed", func() {
+		It("soft reboot vmi with ACPI feature enabled should succeed", decorators.Conformance, func() {
```
This is not optimal:
Why do we care about GA? We should explain it with `By`
Why do we sleep for 30 seconds? Flakes?

xpivarc commented on tests/compute/subresources/vmi.go:
```diff
@@ -203,7 +204,7 @@ var _ = compute.SIGDescribe("VirtualMachineInstance subresource", func() {
 		})
 
-		It("soft reboot vmi should fail to soft reboot a paused vmi", func() {
+		It("soft reboot vmi should fail to soft reboot a paused vmi", decorators.Conformance, func() {
```
Unit test.

xpivarc commented on tests/compute/subresources/vmi.go:
```diff
@@ -59,7 +60,7 @@ var _ = compute.SIGDescribe("VirtualMachineInstance subresource", func() {
 			var err error
 			vmi := libvmifact.NewCirros()
-			vm = libvmi.NewVirtualMachine(vmi, libvmi.WithRunning())
+			vm = libvmi.NewVirtualMachine(vmi, libvmi.WithRunStrategy(v1.RunStrategyAlways))
```
The comment is not meant to be on this line, it is generic.

I would say we miss-place these tests. The `subresources` is technical detail but the real meaning or context of these tests is life cycle.

xpivarc commented on tests/compute/subresources/pausing.go:
```diff
@@ -67,7 +68,7 @@ var _ = compute.SIGDescribe("[rfe_id:3064][crit:medium][vendor:cnv-qe@redhat.com
 		})
 
-		It("[test_id:4597]should signal paused state with condition", func() {
+		It("[test_id:4597]should signal paused state with condition", decorators.Conformance, func() {
```
Again I hate the `BeforeEach`, it doesn't bring value. Quit contrary I need to scroll up and back down

xpivarc commented on tests/compute/subresources/pausing.go:
```diff
@@ -84,7 +85,7 @@ var _ = compute.SIGDescribe("[rfe_id:3064][crit:medium][vendor:cnv-qe@redhat.com
 		})
 
-		It("[test_id:3224]should not be paused with a LivenessProbe configured", func() {
+		It("[test_id:3224]should not be paused with a LivenessProbe configured", decorators.Conformance, func() {
```
And here we see why the `BeforeEach` doesn't eve make sense.

xpivarc replied:
Also this is not Conformance `Expect(err).To(MatchError(ContainSubstring("Pausing VMIs with LivenessProbe is currently not supported")))`, sounds like unit test.

xpivarc commented on tests/compute/subresources/pausing.go:
```diff
@@ -273,28 +246,19 @@ var _ = compute.SIGDescribe("[rfe_id:3064][crit:medium][vendor:cnv-qe@redhat.com
 		})
 
-		It("[test_id:7673]should not pause with dry run", func() {
+		It("[test_id:7673][test_id:7674]should not pause and unpause with dry run", func() {
```
Great merge but these should be moved imho if anything. I would probably reconsider if we can test dry-run as part of unit tests as it is only nice to have functionality.

fossedihelm replied:
We can after https://github.com/kubevirt/kubevirt/pull/13052 will be merged

fossedihelm commented:
@xpivarc As per offline discussion, this PR was entirely refactored:
- More e2e to unit test conversion
- Better placement of the tests (vm/vmi lifecycle instead of subresource file following https://github.com/kubevirt/kubevirt/pull/12865#discussion_r1796878536).
- Less vmi setup on BeforeEach block.

PTAL, Thank you!

xpivarc commented:
/retest

fossedihelm commented:
@xpivarc Rebased and changed the check of the test `[test_id:3085]should be stopped successfully`.
Previously I was checking that:
```
Eventually(matcher.ThisVM(vm)).WithTimeout(300 * time.Second).WithPolling(time.Second).Should(And(
	matcher.BeReady(),
	matcher.HaveConditionTrue(v1.VirtualMachinePaused),
))
```
But this is wrong because the VM is not ready when the vmi is paused, because the pod has the readinessGate:
```
func readinessGates() []k8sv1.PodReadinessGate {
	return []k8sv1.PodReadinessGate{
		{
			ConditionType: v1.VirtualMachineUnpaused,
		},
	}
}
```
And it is copied to the vmi and then to the VM.
Switched to:
https://github.com/kubevirt/kubevirt/blob/ef78d2958a1c3dec5f419dba1d0e1cee9c9da923/tests/compute/vm_lifecycle.go#L135

fossedihelm commented:
@xpivarc :ping_pong:

fossedihelm commented:
/cc @jean-edouard 
Can you take a look? Thanks

xpivarc commented on tests/compute/subresources/pausing.go:
```diff
@@ -70,48 +70,6 @@ var _ = compute.SIGDescribe("[rfe_id:3064][crit:medium][vendor:cnv-qe@redhat.com
 			Eventually(matcher.ThisVM(vm)).WithTimeout(300 * time.Second).WithPolling(time.Second).Should(matcher.BeReady())
 		})
 
-		It("[test_id:4598]should signal paused state with condition", func() {
```
While this is unit testable, it is also expected behavior from use POV so please leave a test that demonstrate this

xpivarc commented on tests/compute/subresources/pausing.go:
```diff
@@ -70,48 +70,6 @@ var _ = compute.SIGDescribe("[rfe_id:3064][crit:medium][vendor:cnv-qe@redhat.com
-			Expect(err).To(MatchError(ContainSubstring("VMI is already paused")))
-		})
-
-		It("[test_id:3060]should signal unpaused state with removed condition", func() {
```
Same as above, the two tests should be one...

fossedihelm replied:
Done

xpivarc commented on tests/vmi_lifecycle_test.go:
```diff
@@ -1512,7 +1512,7 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
 		}
 
-		It("[test_id:7479] should succeed", func() {
+		It("[test_id:7479] should succeed", decorators.Conformance, func() {
```
7480 seem a good candidate and this test seems a subset of it

fossedihelm replied:
Done

fossedihelm commented on tests/subresource_api_test.go:
```diff
@@ -158,49 +158,6 @@ var _ = Describe("[sig-compute]Subresource Api", decorators.SigCompute, func() {
 
 	Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:component] VirtualMachine subresource", func() {
 
-		It("[test_id:1529]should start a VirtualMachine only once", func() {
```
done

fossedihelm commented on tests/subresource_api_test.go:
```diff
@@ -158,49 +158,6 @@ var _ = Describe("[sig-compute]Subresource Api", decorators.SigCompute, func() {
-			Expect(err).To(MatchError(ContainSubstring("VM is already running")))
-		})
-
-		It("[test_id:1530]should stop a VirtualMachine only once", func() {
```
done

fossedihelm commented:
@xpivarc Thank you! 
Comments addressed, PTAL

jean-edouard commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -445,24 +446,31 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 					},
 				}
 
 				vmClient.EXPECT().Get(context.Background(), testVMName, k8smetav1.GetOptions{}).Return(&vm, nil)
 
+				if runStrategy != nil && *runStrategy == v1.RunStrategyManual {
+					vmiClient.EXPECT().Get(context.Background(), testVMName, k8smetav1.GetOptions{}).Return(nil, errors.NewNotFound(v1.Resource("virtualmachineinstance"), vm.Name))
+				}
```
```suggestion
				if runStrategy != nil && *runStrategy == v1.RunStrategyManual {
					vmiClient.EXPECT().Get(context.Background(), testVMName, k8smetav1.GetOptions{}).Return(nil, errors.NewNotFound(v1.Resource("virtualmachineinstance"), vm.Name))
				} else {
					vmClient.EXPECT().Get(context.Background(), testVMName, k8smetav1.GetOptions{}).Return(&vm, nil)
				}
```

jean-edouard reviewed (commented):
/approve

0xFelix commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -2271,31 +2279,49 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 			Expect(response.StatusCode()).To(Equal(http.StatusOK))
-
+			//In case of dry-run the request is not propagated to the handler
+			Expect(backend.ReceivedRequests()).To(handlerExpectation)
```
```suggestion
			Expect(backend.ReceivedRequests()).To(matchExpectation)
```

It's just a rename but IMO that would read more nicely.

0xFelix commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -2271,31 +2279,49 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 
-			expectVMI(running, paused)
+		DescribeTable("Should fail pausing", func(running bool, paused bool, additionalOpts func(vmi *v1.VirtualMachineInstance), pauseOptions *v1.PauseOptions, expectedCode int, expectedError string) {
+			expectVMI(running, paused, additionalOpts)
```
If you make additionalOpts variadic you should be able to get away without adding `nilAdditionalOps`? The third arg on `expectVMI` is already variadic.

fossedihelm replied:
Yes, but I like to have parameters related to the same context close together.
Actually, probably there are too many params here, and this could be improved, but I would leave it for a follow-up. WDYT?

0xFelix replied:
Sounds like a plan.

0xFelix commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -2329,10 +2356,11 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 			Expect(response.StatusCode()).To(Equal(http.StatusOK))
-
+			//In case of dry-run the request is not propagated to the handler
+			Expect(backend.ReceivedRequests()).To(handlerExpectation)
```
Same as above

0xFelix commented on tests/vmi_lifecycle_test.go:
```diff
@@ -1483,6 +1485,95 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
+			Expect(err).ToNot(HaveOccurred())
+			libwait.WaitForSuccessfulVMIStart(vmi, libwait.WithTimeout(180))
+
+			expectedErr := "Internal error occurred"
```
Shouldn't we keep some of the `BeforeEach`s? Also yeah, `expectedErr` looks strange.

fossedihelm replied:
In the end, I opted for doing it here:
[`e955492` (#12865)](https://github.com/kubevirt/kubevirt/pull/12865/commits/e95549209a8f84d34c6b6caf05e3450ae2196347#diff-d9842d3f25d0c190c12e603a60b0af75d5f52f6704866e9db63aeba68bd414f6R1489-R1493)

0xFelix commented on tests/compute/subresources/pausing.go:
```diff
@@ -17,7 +17,7 @@
  */
 
-package tests_test
+package compute
```
Shouldn't this be `subresources`?

0xFelix commented on tests/vmi_lifecycle_test.go:
```diff
@@ -1477,6 +1482,245 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(MatchRegexp("Internal error occurred:.*command Freeze failed:.*QEMU guest agent is not connected"))
+			err = kubevirt.Client().VirtualMachineInstance(testsuite.GetTestNamespace(vmi)).Unfreeze(context.Background(), vmi.Name)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(MatchRegexp("Internal error occurred:.*command Unfreeze failed:.*QEMU guest agent is not connected"))
```
```suggestion
			Expect(err).To(MatchError(MatchRegexp("Internal error occurred:.*command Unfreeze failed:.*QEMU guest agent is not connected")))
```

0xFelix replied:
Should work

fossedihelm replied:
Done :)

0xFelix reviewed (approved):
/lgtm
